kbfstool/md_check.go

```diff
@@ -101,27 +101,124 @@ func checkFileBlock(ctx context.Context, config libkbfs.Config,
 	return nil
 }
 
-func mdCheckOne(ctx context.Context, config libkbfs.Config,
-	input string, rmd libkbfs.ImmutableRootMetadata, verbose bool) error {
-	data := rmd.Data()
+// mdCheckChain checks that the given MD object is a valid successor
+// of the previous revision, and so on all the way back to the given
+// revision. Along the way, it also checks that the root blocks that
+// haven't been garbage-collected are present. It returns a list of MD
+// objects with valid roots, in reverse revision order. If multiple MD
+// objects have the same root (which are assumed to all be adjacent),
+// the most recent one is returned.
+func mdCheckChain(ctx context.Context, config libkbfs.Config,
+	irmd libkbfs.ImmutableRootMetadata,
+	minRevision libkbfs.MetadataRevision, verbose bool) (
+	irmdsWithRoots []libkbfs.ImmutableRootMetadata, err error) {
+	fmt.Printf("Checking chain from rev %d to %d...\n",
+		minRevision, irmd.Revision())
+	gcUnrefs := make(map[libkbfs.BlockRef]bool)
+	for irmd.Revision() > minRevision {
+		rootPtr := irmd.Data().Dir.BlockPointer
+		if !rootPtr.Ref().IsValid() {
+			fmt.Printf("Skipping checking root for rev %d (is invalid)\n",
+				irmd.Revision())
+		} else if gcUnrefs[rootPtr.Ref()] {
+			if verbose {
+				fmt.Printf("Skipping checking root for rev %d (GCed)\n",
+					irmd.Revision())
+			}
+		} else {
+			fmt.Printf("Checking root for rev %d (%s)...\n",
+				irmd.Revision(), rootPtr.Ref())
+			var dirBlock libkbfs.DirBlock
+			err := config.BlockOps().Get(
+				ctx, irmd, rootPtr, &dirBlock)
+			if err != nil {
+				fmt.Printf("Got error while checking root "+
+					"for rev %d: %v\n",
+					irmd.Revision(), err)
+			} else if len(irmdsWithRoots) == 0 ||
+				irmdsWithRoots[len(irmdsWithRoots)-1].Data().Dir.BlockPointer != rootPtr {
+				irmdsWithRoots = append(irmdsWithRoots, irmd)
+			}
+		}
+
+		for _, op := range irmd.Data().Changes.Ops {
+			if gcOp, ok := op.(*libkbfs.GCOp); ok {
+				for _, unref := range gcOp.Unrefs() {
+					gcUnrefs[unref.Ref()] = true
```

Comment:
`GCOp` contains the revision number that was last gc'd -- you don't really have to check blocks like this.  But I could see an argument for this way being a stronger protection against broken GC, so whichever way you prefer is fine.